### DISCUSSION TITLE
mDNS support

### DISCUSSION THREAD

**Discussion by phoddie** [phoddie/node-red-mcu#113] [Ideas]
I'm a big fan of mDNS. It solves a bunch of problems for connecting devices together on a local network. The protocol design is both elegant and efficient. The Moddable SDK has an [implementation of mDNS](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/network/network.md#mdns) that supports claiming local network names, service discovery, and service announcements.

The OTA Update work and a recent question from @bartbutenaers made me think about mDNS. For example, to Push an OTA Update you need the IP address of the device. With mDNS, you would just use the name, like `mysensor.local`. You could get fancier and announce an OTA Update service with mDNS. That would allow tools (like the MCU plugin) to discover devices on the network that accept OTA Updates, display them in a list, and send updates directly.

It seems like this would be pretty useful for Node-RED, but I don't see that much use of mDNS in the Node-RED universe. The definitive mDNS Node seems to be the [node-red-node-discovery](https://flows.nodered.org/node/node-red-node-discovery) node included in the venerable "node-red-nodes" repository. That supports service announcements and discovery.

But, the node-red-node-discovery node doesn't support claiming a name. I suppose that's because it runs on a Raspberry Pi and there are mechanisms used in Linux to claim mDNS names, so it isn't part of Node-RED?

Anyway... before taking time on implementation, I'd appreciate any feedback on this direction. For example, maybe there's a good reason I'm overlooking that mDNS isn't used much with Node-RED.  Thanks!

**Comment by bartbutenaers** [comment]
@phoddie,
Unfortunately I don't have much feedback about this.  If you think this might be useful, it might be better to start a discussion on Discourse.  But if you need to spend a lot of time on it, just for my question then please skip my question!

I was just wondering if I could somehow (e.g. via the xsbug-log logging see the MAC or IP address of my MCU).  That would be useful because on my router's DHCP service, I always map IP address to the MAC addresses:

![image](https://user-images.githubusercontent.com/14224149/232891098-46f53afb-b914-4bee-bc73-331381129ccf.png)

Since my DHCP server passes always the same IP address to a device (based on its MAC address), I can use these (fixed) IP addresses in Node-RED to communicate to my devices (without specifying a fixed IP addres in each of my devices individually).  But it is not easy to see in my current router's user interface to see new devices popping up, so it would be useful I could see the MCU's IP or MAC address somewhere in the xsbug-log logging.  Perhaps I just overlooked it somewhere...

Bart

**Reply by phoddie** [reply]
Regarding mDNS, given that it isn't used much at all in Node-RED (for computers), I doubt the Discourse will have much to say. But, I can try. More to the point, even if it is useless for Node-RED on computers, it could still be valuable on MCUs. That's why I posted here first.

The IP address is displayed when you start up. Since you don't seem to be seeing it, here's an image from the MCU plug-in sidebar to be sure we are talking about the same thing:

<img width=450 src="https://user-images.githubusercontent.com/6856458/232892635-2b48cc54-6766-42e3-b6c1-5ecf64ca7178.png"/>

Since this is Node-RED, you can gather all this network status information in a flow too. Here's a flow that is the equivalent of the Moddable SDK's [Net module example](https://github.com/Moddable-OpenSource/moddable/blob/public/examples/network/net/main.js). It uses a Function node to access the Net module.

<details>

```json
[
    {
        "id": "8dbc54f41c619c50",
        "type": "tab",
        "label": "demo",
        "disabled": false,
        "info": "",
        "env": [],
        "_mcu": {
            "mcu": true
        }
    },
    {
        "id": "6cb441c35f968994",
        "type": "debug",
        "z": "8dbc54f41c619c50",
        "name": "",
        "active": true,
        "tosidebar": true,
        "console": true,
        "tostatus": false,
        "complete": "payload",
        "targetType": "msg",
        "statusVal": "",
        "statusType": "auto",
        "_mcu": {
            "mcu": true
        },
        "x": 460,
        "y": 180,
        "wires": []
    },
    {
        "id": "66aa43e34872d760",
        "type": "function",
        "z": "8dbc54f41c619c50",
        "name": "network status",
        "func": "msg.payload = {\n    IP: Net.get(\"IP\"),\n    MAC: Net.get(\"MAC\"),\n    SSID: Net.get(\"SSID\"),\n    BSSID: Net.get(\"BSSID\"),\n    RSSI: Net.get(\"RSSI\")\n};\n\nreturn msg;\n",
        "outputs": 1,
        "noerr": 0,
        "initialize": "",
        "finalize": "",
        "libs": [
            {
                "var": "Net",
                "module": "net"
            }
        ],
        "_mcu": {
            "mcu": true
        },
        "x": 260,
        "y": 180,
        "wires": [
            [
                "6cb441c35f968994"
            ]
        ]
    },
    {
        "id": "1a5cddab831ae9ac",
        "type": "inject",
        "z": "8dbc54f41c619c50",
        "name": "once",
        "props": [],
        "repeat": "",
        "crontab": "",
        "once": true,
        "onceDelay": 0.1,
        "topic": "",
        "_mcu": {
            "mcu": true
        },
        "x": 110,
        "y": 180,
        "wires": [
            [
                "66aa43e34872d760"
            ]
        ]
    }
]
```

</details>

Here's what the output looks like from an ESP32 at the Moddable office.

<img width="700" alt="image" src="https://user-images.githubusercontent.com/6856458/232893940-57b59b2e-0b73-45b4-9377-805851fdfc81.png">

**Comment by colinl** [comment]
Also, with the ESP8266, when I run mcconfig and it downloads the code to the device I see
```
esptool.py v2.4.0
Connecting....
Detecting chip type... ESP8266
Chip is ESP8266EX
Features: WiFi
MAC: 84:cc:a8:a4:50:bf
...
```
